### PR TITLE
Fix #28 -- escaping arguments as mentioned in TODO

### LINKED ISSUE #28: Implement proper shell escaping for pub command error reproduction
<a href="https://github.com/whesse"></img></a> **Issue by [whesse](https://github.com/whesse)**
_Originally opened as dart-lang/sdk#23505_

---

Error reporting from pub includes printing a command line to rerun the failing command.  The arguments need to be properly escaped, depending on the shell (OS) that is running.  We currently put double quotes around any argument containing spaces.

### PR DESCRIPTION
**Pull request by jonasfj**
I just want to close: https://github.com/dart-lang/pub/issues/28

### CODE REVIEW

File: lib/src/command_runner.dart [hunk 1]
```diff
@@ -180,8 +180,11 @@ class PubCommandRunner extends CommandRunner {
       if (topLevelResults['trace']) {
         log.dumpTranscript();
       } else if (!isUserFacingException(error)) {
-        // TODO(23505): Implement proper shell escaping, not a partial hack.
-        protectArgument(String x) => x.contains(' ') ? '"$x"' : x;
+        // Escape the argument for users to copy-paste in bash.
+        protectArgument(String x) =>
+            RegExp(r'^[a-zA-Z0-9-_]+$').stringMatch(x) == null
+                ? "'${x.replaceAll("'", "'\\''")}'"
```
Comment (sigurdm) [inline lib/src/command_runner.dart]: You cannot escape a single quote inside a single quote:
https://www.gnu.org/software/bash/manual/html_node/Single-Quotes.html#Single-Quotes
You probably need to use double quotes and escape all of ‘$’, ‘`’, ‘\’

Comment (jonasfj) [reply lib/src/command_runner.dart]: Not really... Read carefully I escape single quote, by ending single quotation then having an escaped single quote and starting single quotation again :)

If there is no space between end and start of single quotation it doesn't become a new argument...

Comment (sigurdm) [reply lib/src/command_runner.dart]: Explain this in a comment ...

Comment (sigurdm) [inline lib/src/command_runner.dart]: You could use `r"'\''"`

**Pull request merged**